atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/ConflictHandler.java

```diff
@@ -60,13 +60,76 @@
      * This should not be used on tables that are expected to have tons of reads because we must keep the whole read
      * set in memory so we can compare it at commit time to ensure that everything we read has not changed.
      */
-    SERIALIZABLE,
+    SERIALIZABLE(false, true, true, true),
 
     /**
      * Same as {@link RETRY_ON_WRITE_WRITE}, but checks for conflicts by locking cells during commit instead
      * of locking rows. Cell locks are more fine-grained, so this will produce less contention at the expense
      * of requiring more locks to be acquired.
      */
-    RETRY_ON_WRITE_WRITE_CELL;
+    RETRY_ON_WRITE_WRITE_CELL(true, false, true, false),
 
+    /**
+     * Same as {@link SERIALIZABLE}, but checks for conflicts by locking cells during commit instead
+     * of locking rows. Cell locks are more fine-grained, so this will produce less contention at the expense
+     * of requiring more locks to be acquired.
+     */
+    SERIALIZABLE_CELL(true, false, true, true),
+
+    /**
+     * This conflict handler is designed to be used by index tables. As any write/write conflict on an index table
+     * will necessarily also be a write/write conflict on base table, we do not need to check write/write conflicts
+     * on index tables. Read/write conflicts should still need to be checked, since we do not need to read the index
+     * table with the main table.
+     * <p>
+     * This conflict handler locks at cell level. As index tables use dynamic columns, locking at row level would
+     * result in higher contention.
+     */
+    SERIALIZABLE_INDEX(true, false, false, true),
+
+    /**
+     * This conflict handler is designed to be used for migrating a table from SERIALIZABLE to SERIALIZABLE_CELL
+     * conflict handler, or vice versa without requiring a shutdown.
+     * <p>
+     * Current implementation of lock service permits a client to lock a cell after locking the row that cell
+     * belongs to. In a scenario where we migrate table A to SERIALIZABLE_CELL from SERIALIZABLE, and conduct a rolling
+     * upgrade, if client_1 is on the newer version, it will try to acquire cell locks for its transaction t1; while
+     * client_2 with an older version, acquiring row locks for its transaction t2. If t1 and t2 are modifying the same
+     * cell then correctness issues may occur as there is no way for t1 to be aware of t2. Locking on both cell and row
+     * level prevents this issue.
```

Comment:
When can clients safely migrate from `SERIALIZABLE_LOCK_LEVEL_MIGRATION` to `SERIALIZABLE_CELL`?

Reply:
when all nodes are running with one, or the other

Reply:
it's safe to run S_C with S_L_L_M and it's safe to run S_L_L_M with S, but not S_C with S. So you have a quick online migration for the change.